Mx ocrmypdf

mimimi1968 opened the pull request:
First: I like the project and it was the point for me to really switch over to a paperless workflow, thanks a lot for your work!
Second: I'm new to python and django. I used to code in C and C++, but python always stood on the todo-list... So this are my first lines ever in python. And django? It's really really cool but a lot to learn.

So in short I made the following changes:
* fight against python, pip, docker, alpine, pypi, dependencies... ;-)
* added a ocrmypdf based parser, because I wanted the OCR to put the text into my PDFs. Unfortunately this feature was sold separately and only for Windows in addition to my scanner (Brother ADS 1700W)
* changed the way, the consumer works: it can "eat" the input documents (default) and it can preserve the unchanged input files. So I can put the originals again into some more advanced OCR tool just in case in the future.
* the consumer had some problems due to the slow transfer from the scanner into the consume dir together with docker on MacOS: the no-inotify method ended up with zero length documents and errors
* i18n for most of the admin GUI
* some more small enhancements like actions for tags and correspondents directly in the form

I hope someone will find it useful.

mimimi1968 committed:
Switch to ocrmypdf

mimimi1968 committed:
Introduce 'move' feature in infolder

mimimi1968 committed:
Exclude ./consume dir

mimimi1968 committed:
Optimize order in Dockerfile

mimimi1968 committed:
Catch the exception when Ocrmypdf dies

mimimi1968 committed:
Reworked the consume mechanism to work better

This change was needed for slowly growing files coming in from
scanner. The old implementation grabbed them even when they were
not completely written.
So we wait now (when not using inotify) and watch for changes in
mtime and/or size of the file. Normally this gives enough time to
the scanning process to complete the work and transfer.

mimimi1968 committed:
Made LANGUAGE configurable

mimimi1968 committed:
Just corrected the path for static/

mimimi1968 committed:
Include some management actions into GUI

mimimi1968 committed:
Renamed gunicorn.conf to remove warning

mimimi1968 committed:
Updated Django to 2.2 and fixed the deps

mimimi1968 committed:
Started i18n localization

mimimi1968 committed:
Filter for "added" field too

mimimi1968 committed:
Remove hack for date input

mimimi1968 committed:
Corrected the Dockerfile after squashing the commits

mimimi1968 committed:
Upgraded to alpine 3.12 - it just worked

mimimi1968 committed:
Fixed the tests to pass at least

mimimi1968 committed:
Finished the i18n for admin interface

mimimi1968 committed:
Added the forgotten package dep

mimimi1968 committed:
Added jquery.min.js so the colouring works now

mimimi1968 committed:
Changed some translated words to be consistent

mimimi1968 committed:
Rollback to alpine 3.10 for tesseract not working in 3.12

The ocrmypdf process came back with tesseract took too long...
So instead investigating this, I took the shortcut.

mimimi1968 committed:
Raised the worker timeout

It was too short when deleting som 10 files. So I set this to 90s
for now.

mimimi1968 committed:
Fixed the whitespace handling and small improvement for get_text

mimimi1968 committed:
Better handling of change detection when scanning /consume

mimimi1968 committed:
Add a page count to gui and business logic

mimimi1968 committed:
Changed the 'created' Field to DateField for
better user experience

mimimi1968 committed:
Display the version information in footer

mimimi1968 committed:
Ignore further changes of version.txt

mimimi1968 committed:
Silence pycodestyle errors and warnings

mimimi1968 committed:
Instruct ocrmypdf to rotate when needed

mimimi1968 committed:
Enable forcing the conversion again

mimimi1968 committed:
Update to Alpine 3.11

mimimi1968 committed:
Move transaction to _store function

mimimi1968 committed:
Silence optipng output

mimimi1968 committed:
Make ocrmypdf more robust and debug friendly

mimimi1968 committed:
Remove build for python 3.5 and add 3.8

mimimi1968 committed:
Just added some .keep files

mimimi1968 committed:
Keep the fiel metadata when exporting/importing

mimimi1968 committed:
Added the field "pages" to the serialiser code

mimimi1968 committed:
Removed the misleading error message for now

The handling of the source filename is a kind of non trivial,
possibly broken. So this is work for a later moment...
When the file object is created, the file doesn't have to exist,
but this is assumed for now (and searched for a moved file).

mimimi1968 committed:
Refactored the directory/file handling

There were different functions to cleanup the directory structure
in MEDIAROOT/originals when using a intelligent file naming format.
This was reduced and cleaned up to be more to the point. The idea is
to address all situations, when a file is moved or could be renamed
by us or others. After this it is time to clean up empty dirs.
There is no need for optimization, so one recursive function can do
this for us.

mimimi1968 committed:
Create directories in MEDIAROOT when we save a new document

mimimi1968 committed:
Updated the packages

mimimi1968 committed:
Support for mysql as database

mimimi1968 committed:
Update packages because of factoryboy not running

pitkley commented:
Thank you for this huge contribution! I just wanted to let you know that you will have a very hard time getting a review for this PR, mainly because it is incredibly large and contains multiple, distinct features, which makes reviewing it extra tough.

If you want to see your changes land in Paperless, your best bet is to split this up into separate PRs, where each PR contains a distinct feature.

Also, without looking into this in detail, dropping support for Python 3.5, while sensible since it is EOL, at least requires an update to the documentation, which I'm not seeing. I'm not seeing any documentation at all, and for new features it would be nice to at least have short descriptions for what each thing does (and what might change for users of the application, because there seem to be quite a few changes that could be breaking to our users).

mimimi1968 committed:
Fixed an i18n issue in change form

mimimi1968 commented:
Thank you for your nice notice and the hints! I made some development straight forward to suit my needs. Now the current state is some point, where I reason what to do next. I can break down the rather huge change into small and distinct branches, each a separate PR. But I'm unsure if this is the way for me. Yes it is completely understood, that any merge into upstream needs review. My main question is, will there be any review in the next future?

Second point of interest is if my changes are acceptable at all? I tried not to break the behavior, beside some changes e.g. in the consumer code, where I let the configuration decide. So I made some rather small enhancements, started i18n with german localization, fixed small errors. 

One big feature was the support for another OCR backend, that indeed changes the content of the consumed document to have a searchable PDF.
Is there "a big plan" for features, acceptable changes or some guidance for such questions?
Of cause I owe some updated documentation. I will add this soon, at least for me to remember the design decisions.

Thanks again for the time you spent looking into it and writing the comment.

mimimi1968 committed:
Introduced an upload facility for documents

mimimi1968 committed:
Silence pycodestyle warnings

mimimi1968 committed:
Fix some C-style format strings

mimimi1968 committed:
Add some built-time deps for Pillow package

mimimi1968 committed:
Fixes again for the last 2 commits

mimimi1968 committed:
Switch over to a binary dependency on ocrmypdf

It was too hard to build a reliably working ocrmypdf. So for
now we use the pre-built docker container and make a multi-stage
build for the rest.

mimimi1968 committed:
Add /usr/local/bin to find ocrmypdf in path

mimimi1968 committed:
Silence a db warning for mysql backend

mimimi1968 committed:
Optimized Dockerfile for size of the image

mimimi1968 committed:
Lock the dependencies for now
